### PR TITLE
try-runtime-cli: parallel key scraping

### PR DESCRIPTION
**Pull request by surajk-m**
This PR implements try-runtime-cli: parallel key scraping process. Fixes #__14085__

The implementation divides the prefix into `NUM_PARTS` parts and creates a separate `rpc_get_keys_paged_part` task for each part, which is executed concurrently using `join_all`. Each task fetches keys from its own range of the prefix and returns them to the main task, which aggregates the keys from all the parts and filters out default child storage.

We should select the value of `N` (the number of parts) carefully to optimize the balance between parallelism and efficiency. If we choose a value of `N` that is too large, we may end up creating too many parallel tasks, which could result in additional overhead and slow down the overall process. On the other hand, if we choose a value of `N` that is too small, we may not be able to fully utilize the available resources, resulting in slower overall performance.

### CODE REVIEW

File: utils/frame/remote-externalities/src/lib.rs [hunk 1]
```diff
@@ -351,9 +351,53 @@ where
...
 	) -> Result<Vec<StorageKey>, &'static str> {
-		let mut last_key: Option<StorageKey> = None;
+		const NUM_PARTS: usize = 4; // number of parts to divide the prefix into
+
+		let mut tasks = Vec::with_capacity(NUM_PARTS);
+
+		// Divide the prefix into N parts
+		let prefix_bytes = prefix.0;
+		let total_len = prefix_bytes.len();
+		let part_len = total_len / NUM_PARTS;
```
Comment (ggwpez) [inline utils/frame/remote-externalities/src/lib.rs:361]: I dont get how this is supposed to work.  
Splitting a prefix happens by appending to it, not cutting it appart.

Comment (s5k0651) [reply utils/frame/remote-externalities/src/lib.rs:361]: I'm dividing the byte representation of the prefix into N parts by splitting it into equal parts of length `part_len`. This is not the same as appending to the prefix, as you correctly pointed out.

Can I split the prefix into four parts by changing the first byte of the key. Each part has a size of 256, which is the size of a byte. The start key of each range is set to the original prefix, and the end key is set by incrementing the first byte of the start key by 1.

File: utils/frame/remote-externalities/src/lib.rs [hunk 2]
```diff
@@ -351,14 +351,57 @@ where
...
+		let mut tasks = Vec::with_capacity(NUM_PARTS);
+
+		// Divide the prefix into N parts
+		let mut start_key = prefix.clone();
+		let mut end_key = prefix.clone();
 		let mut all_keys: Vec<StorageKey> = vec![];
-		let keys = loop {
+
+		for _ in 0..NUM_PARTS {
+			end_key.0[0] = start_key.0[0].wrapping_add(1);
```
Comment (liamaharon) [inline utils/frame/remote-externalities/src/lib.rs:364]: This line panics when run against Polkadot state:

```
$ cargo run --features=try-runtime try-runtime --runtime ../polkadot/target/release/wbuild/polkadot-runtime/polkadot_runtime.wasm on-runtime-upgrade live --uri ws://localhost:9944
...
Thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 0', /Users/liamaharon/grimoire/substrate/utils/frame/remote-externalities/src/lib.rs:364
```

How're you testing it?

File: utils/frame/remote-externalities/src/lib.rs [hunk 3]
```diff
@@ -351,14 +351,57 @@ where
...
+		for task in futures::future::join_all(tasks).await {
+			let mut keys = task?;
+			all_keys.append(&mut keys);
+		}
+
+		// Filter out default child storage keys
+		let filtered_keys = all_keys
+			.into_iter()
+			.filter(|k| !is_default_child_storage_key(&k.0))
+			.collect::<Vec<_>>();
```
Comment (liamaharon) [inline utils/frame/remote-externalities/src/lib.rs:379]: why'd you add filtering here?

File: utils/frame/remote-externalities/src/lib.rs [hunk 4]
```diff
@@ -351,14 +351,57 @@ where
...
+			end_key.0[0] = start_key.0[0].wrapping_add(1);
+			let task = self.rpc_get_keys_paged_range(start_key.clone(), end_key.clone(), at);
+			tasks.push(task);
+			start_key.0[0] = end_key.0[0].wrapping_add(1);
+		}
+
+		for task in futures::future::join_all(tasks).await {
+			let mut keys = task?;
+			all_keys.append(&mut keys);
+		}
```
Comment (liamaharon) [inline utils/frame/remote-externalities/src/lib.rs:373]: I'd prefer if this feature was implemented with a more functional approach rather than procedural, similar to https://github.com/paritytech/substrate/blob/ab8330cd421027d0ebfd613111702fd4fa5e9487/utils/frame/remote-externalities/src/lib.rs#L560-L578

I feel functional is easier to follow and less bug-prone

File: utils/frame/remote-externalities/src/lib.rs [hunk 5]
```diff
@@ -351,14 +351,57 @@ where
...
+		let mut end_key = prefix.clone();
 		let mut all_keys: Vec<StorageKey> = vec![];
-		let keys = loop {
+
+		for _ in 0..NUM_PARTS {
+			end_key.0[0] = start_key.0[0].wrapping_add(1);
+			let task = self.rpc_get_keys_paged_range(start_key.clone(), end_key.clone(), at);
+			tasks.push(task);
+			start_key.0[0] = end_key.0[0].wrapping_add(1);
+		}
```
Comment (liamaharon) [inline utils/frame/remote-externalities/src/lib.rs:368]: it's not clear to me how this splits the prefix into N equal sized parts and appears to be buggy with the 'index out of bounds' error.

to make this code more solid, I suggest:
1. abstract the prefix splitting code into a dedicated function
2. add comments into the function clearly explaining how it works (including examples in the comments if you can) and 
3. add a test for the function making sure it handles all regular and edge cases you can think of

**Review by liamaharon — COMMENTED** [review]
Thanks for having a go at this, added some comments

File: utils/frame/remote-externalities/src/lib.rs (hunk 2 above)
Comment (s5k0651) [reply utils/frame/remote-externalities/src/lib.rs]: Earlier I had only build the node with --features try-runtime flag in release mode. Thanks for the comments, I've updated the implementation.

File: utils/frame/remote-externalities/src/lib.rs (hunk 3 above)
Comment (s5k0651) [reply utils/frame/remote-externalities/src/lib.rs]: Yes, It's not required. I completely overlooked the rpc_get_pairs_paged function.

File: utils/frame/remote-externalities/src/lib.rs (hunk 5 above)
Comment (s5k0651) [reply utils/frame/remote-externalities/src/lib.rs]: I've abstracted the prefix splitting code in a separate function named `split_storage_key_prefix`  and have also added comments explaining how it works.

File: utils/frame/remote-externalities/src/lib.rs [hunk 6]
```diff
@@ -350,15 +350,90 @@ where
...
+	) -> Vec<(StorageKey, StorageKey)> {
+		let mut start_key = prefix.clone();
+		let mut chunks = vec![];
+
+		// loops `num_parts` times, each time calling the `get_next_wrapping_index` function
+		// with the current `start_key` value, and pushing the result as a tuple
+		// of `(start_key, end_key)` onto the `chunks` vector. The `start_key` value is then
+		// updated to the `end_key` value
+		for _ in 0..num_parts {
+			let end_key = StorageKey(Self::get_next_wrapping_index(&start_key.0));
```
Comment (liamaharon) [inline utils/frame/remote-externalities/src/lib.rs:389]: How does this know what `end_key` should be without knowing how many chunks there are? Does incrementing each byte by one really split the key into equal sized chunks?

I'm still generally confused by how this splitting logic works. Could you please write a test showing examples of it splitting keys into equal sized chunks in all sorts of different scenarios?

Comment (s5k0651) [reply utils/frame/remote-externalities/src/lib.rs:389]: Sure, I'll write the test cases for the `split_storage_key_prefix` logic with some different scenarios.

The splitting logic works by dividing the range between the `prefix` and the maximum possible key value into equal-sized chunks, represented as pairs of `(start_key, end_key)`, where `start_key` is the starting key of the chunk, and `end_key` is the ending key of the chunk.

The number of chunks is determined by the `NUM_PARTS` constant, it is divided into 4 equal-sized chunks. The splitting logic does not know the exact number of chunks beforehand, but it divides the range into equal-sized chunks based on the `NUM_PARTS` constant. The `get_next_wrapping_index` function is used to calculate the `end_key` of each chunk, and the `start_key` of each chunk is simply the `end_key` of the previous chunk.

The  `get_next_wrapping_index` function is used to calculate the `next wrapping index` of a given byte slice key. By incrementing each byte by one, it returns a new byte slice that represents the next key in sequence.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2899214

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2899213

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2899392

**Review by liamaharon — CHANGES_REQUESTED** [review]
Thanks for adding these tests, they'll help a lot getting this over the line.

I see three areas for improvement: 

1. A `StorageKey` is a base-16, meaning we need to cover `0` through to `15` (`0` -> `f` in hex) to make sure we cover the entire prefix. The distance between each tuple of keys should be as equal as possible, while covering `0` through `15`.
2. The prefix needs to remain the same on your output keys, notice I added the extra zero in my example below. Currently you're actually iterating over the entire `0,0,0` prefix just in `(StorageKey(vec![0, 0, 0]), StorageKey(vec![0, 0, 1])),`.
3. It probably needs a check to make sure the function won't spit out invalid storage keys if a prefix is passed that cannot be subdivided further.

Example of 1, 2 changes applied to your num_parts = 5 test: the test should output something more like this:

```rust
                // prefix is 0000. so we want to cover 0000a,0000b,...,0000e,0000f.
		assert_eq!(
			result,
			vec![
                                 // covers 00000 - 00002
				(StorageKey(vec![0, 0, 0, 0, 0]), StorageKey(vec![0, 0, 0, 0, 2])),    
                                 // covers 00003 - 00005
				(StorageKey(vec![0, 0, 0, 0, 3]), StorageKey(vec![0, 0, 0, 0, 5])),
                                 // covers 00006 - 00008   
				(StorageKey(vec![0, 0, 0, 0, 6]), StorageKey(vec![0, 0, 0, 0, 8])),    
                                 // covers 00009 - 0000b
				(StorageKey(vec![0, 0, 0, 0, 9]), StorageKey(vec![0, 0, 0, 0, 11])),   
                                 // covers 0000c - 0000f
				(StorageKey(vec![0, 0, 0, 0, 12]), StorageKey(vec![0, 0, 0, 0, 15])), 
			]
		);
```

Please let me know if anything in unclear.

**Comment by s5k0651** [comment]
Thanks for the comments @liamaharon, I'm working on the PR with the improvements.
